### ISSUE TITLE
No License Specified / Exception Handling

### ISSUE THREAD

**Issue by dahvid** [n-s-f/taskmap#2]
Hi,
I'd appreciate if you'd specify a license.  No license is very ambiguous and I'd like to use your code in a research project.
Thanks,
David

**Comment by n-s-f** [comment]
Yep, done. Can I ask what research project you'll be using it for?

**Comment by dahvid** [comment]
Hi Noam,
I work at GE research on DL stuff.  I'll be using it for an in house DL
tool we have.  Taskmap isn't really crucial but I think it will save me
some time. I really like how simple and easy to comprehend it is.
Regards,
David

**Comment by n-s-f** [comment]
Great, hope it's helpful!

**Comment by dahvid** [comment]
Hi Noam,
I see your run_task() function eats exceptions.  It logs, but not the
entire stack-trace. This makes debugging complex tasks quite difficult.
I'd like to add the ability to turn off exception eating and (when eating
is on) log the stack trace as well.  I don't want to mess up the philosophy
of your work. What would be a good way to do this?
Regards,
David

**Comment by n-s-f** [comment]
Hi David,

Thanks for the message.

The logging issue can be addressed in the `task_error` function.  Instead of logging the exception on line [48](https://github.com/n-s-f/taskmap/blob/master/taskmap/taskmap.py#L48), log the whole stack trace.  I don't have time to make this change, but feel free to make a pull request. I think it's a great idea.

I'm not exactly sure what the other issue is. Do you mean you want the whole pipeline to stop running as soon as one task fails? If so we could make that configurable by passing an option to the graph like `stop_on_error`, which would get checked in the [run_task](https://github.com/n-s-f/taskmap/blob/master/taskmap/taskmap.py#L18-L26) and [run_task_async](https://github.com/n-s-f/taskmap/blob/master/taskmap/taskmap.py#L29-L37) functions.

That said I don't really like this idea. Why not test the tasks in isolation, or run a smaller subgraph if there are dependencies? Can you elaborate on when this might be needed?

It might also be relevant to point this function to help rerun only the failed tasks if you haven't seen it yet.
https://github.com/n-s-f/taskmap#taskmapreset_failed_tasks

**Comment by dahvid** [comment]
Hi Noam,
Thanks for the suggestions, that's what I was looking for.  I actually have
a hard time thinking of a use case for re-starting failed tasks.  Unless
we're talking about tasks that are doing network access, they will just
fail again!   If we're talking about network tasks, the semantics required
for fail safety are much more complicated then simply "try again".   I'll
add these features and do a pull request. I have three essential use cases
here, one running on a  single node within the same process space, where I
want an immediate failure, or running in  a Docker container where I want
an immediate exit with a logged stack trace. Remember I am running tasks
that may take a long time to run (Tensorflow training) and no failures are
allowed in the graph execution.  Higher level network failures are being
handled by the distributed system, which is also helped in it's job by
immediate failure of a single process. Maybe we need two options,
stop_on_error and no_handle_exception.
Regards,
David

On Thu, Mar 22, 2018 at 11:23 PM, Noam Finkelstein <notifications@github.com

**Comment by dahvid** [comment]
Hi Noam,
I added the stack trace to the error message.  But... consider the
following things.

1)
Task A - runtime 1 hour
Task B - runtime 1 hour
Task C - depends on both A and B

Run parallel
Task A fails immediately, Task B continues to run.  I wait an hour to find
out I had a silly syntax error.

2)
Why should I write a bunch of code (and attendant bugs) in order to, parse
the messages of unfinished tasks, decide which one failed based on the
message (as opposed to merely not completed due to dependencies) and then
re-throw the exception with the old stack trace.  The interpreter does what
I want automatically if I don't mess with it.

3)
Your library is not an application, but it calls application code.  By
trapping exceptions you've prevented applications from throwing and
handling their own exceptions.  I think this is one reason trapping all
exceptions is considered bad form in Python.

4)
I use pytest for unit tests, I was initially fooled by my code throwing no
exceptions into thinking it was working, it took me a while to figure out
what was going on.  I would need to write extensive code in order to parse
the log files for each unit test.  It's a lot easier to just catch an
exception and assert False, I know exactly where the exception occurred
immediately.

In the meantime I've added a trap_exceptions=True variable, that if set to
False (default is True) that turns off exception trapping.  It's allowed me
to continue to work without a lot of hassle.

Regards,
David
